viewing_party/party.py

```diff
@@ -1,23 +1,139 @@
+    return user_data
+
+
+def add_to_watchlist(user_data, movie):
+    watchlist = dict()
+    watchlist["title"] = movie["title"]
+    watchlist["genre"] = movie["genre"]
+    watchlist["rating"] = movie["rating"]
+    user_data["watchlist"].append(watchlist)
+    return user_data
+
+
+def watch_movie(user_data, title):
+    for m in user_data["watchlist"]:
+        if m["title"] == title:
+            user_data["watchlist"].remove(m)
+            user_data["watched"].append(title)
+            break
+    return user_data
+
 
 # -----------------------------------------
 # ------------- WAVE 2 --------------------
 # -----------------------------------------
+def get_watched_avg_rating(user_data):
+    result = 0
+    count = len(user_data["watched"])
+    for m in user_data["watched"]:
+        result += m["rating"]
+    if count:
+        result /= count
+    return result
+
+
+def get_most_watched_genre(user_data):
+    genre_dict = dict()
+    result = None
+    if user_data["watched"]:
+        for m in user_data["watched"]:
+            if m["genre"] in genre_dict.keys():
+                genre_dict[m["genre"]] += 1
+            else:
+                genre_dict[m["genre"]] = 1
+        # sort dictionary in ascending order
+        result = sorted(genre_dict.items(), key=lambda x: x[1])[-1][0]
+    return result
 
 
 # -----------------------------------------
 # ------------- WAVE 3 --------------------
 # -----------------------------------------
+def get_unique_watched(user_data):
+    user_watched = user_data["watched"]
+    friends_watched = list()
+    for f in user_data["friends"]:
+        for m in f["watched"]:
+            friends_watched.append(m)
+    unique = [i for i in user_watched if i not in friends_watched]
+    return unique
+
+
+def get_friends_unique_watched(user_data):
+    user_watched = user_data["watched"]
+    friends_watched = list()
+    for f in user_data["friends"]:
+        for m in f["watched"]:
+            friends_watched.append(m)
+    # remove duplicates
+    friends_watched_nd = list()
+    [friends_watched_nd.append(i) for i in friends_watched if i not in friends_watched_nd]
+    unique = [i for i in friends_watched_nd if i not in user_watched]
+    return unique
 
         
 # -----------------------------------------
 # ------------- WAVE 4 --------------------
 # -----------------------------------------
+def get_available_recs(user_data):
+    recs = list()
+    friends_watched = list()
+    for f in user_data["friends"]:
+        for m in f["watched"]:
+            friends_watched.append(m)
+    # remove duplicates
+    friends_watched_nd = list()
+    [friends_watched_nd.append(i) for i in friends_watched if i not in friends_watched_nd]
+
+    if (len(user_data["subscriptions"])!=0) and (len(friends_watched)!=0):
+        for i in friends_watched_nd:
+            if i["host"] in user_data["subscriptions"] and i not in user_data["watched"]:
+                recs.append(i)
+    return recs

```

Comment:
It looks like we duplicate some of our prior logic here. Another option would be to reuse the `get_friends_unique_watched` function to get a list of movies our users has not seen, then filter that list by available subscriptions:

```python
unique_friend_movies = get_friends_unique_watched(user_data)
recs = []

for movie in unique_friend_movies:
    if movie['host'] in user_data['subscriptions']:
        recs.append(movie)

return recs
```